pandas/parser.pyx

```diff
@@ -1468,6 +1505,97 @@ cdef _string_box_decode(parser_t *parser, int col,
 
     return result, na_count
 
+@cython.boundscheck(False)
+cdef _categorical_convert(parser_t *parser, int col,
```

Comment:
The thought was it'd be more performant - you already have to hash each element to compare to na, so why create the uniques at that point, and use the much smaller set for type inference.